File: 07. Greedy/06. Insert Interval/main.js
function insert(intervals, newInterval) {
    let result = [];
    let i = 0, n = intervals.length;

    // Before the new interval
    while (i < n && intervals[i][1] < newInterval[0]) {
        result.push(intervals[i]);
        i++;
    }

    // Merge overlapping intervals
    while (i < n && intervals[i][0] <= newInterval[1]) {
        newInterval[0] = Math.min(newInterval[0], intervals[i][0]);
        newInterval[1] = Math.max(newInterval[1], intervals[i][1]);
        i++;
    }
    result.push(newInterval);

    // After the new interval
    while (i < n) {
        result.push(intervals[i]);
        i++;
    }

    return result;
}

// Main function
function main() {
    let intervals = [[1, 3], [6, 9]];
    let newInterval = [2, 5];

    let result = insert(intervals, newInterval);
    console.log("Updated intervals:", result);
}

// Run main function
main();

File: 07. Greedy/06. Insert Interval/main.java
import java.util.*;

public class main {
    public static List<int[]> insert(int[][] intervals, int[] newInterval) {
        List<int[]> result = new ArrayList<>();
        int i = 0, n = intervals.length;

        // Before new interval
        while (i < n && intervals[i][1] < newInterval[0]) {
            result.add(intervals[i]);
            i++;
        }

        // Merge overlapping intervals
        while (i < n && intervals[i][0] <= newInterval[1]) {
            newInterval[0] = Math.min(newInterval[0], intervals[i][0]);
            newInterval[1] = Math.max(newInterval[1], intervals[i][1]);
            i++;
        }
        result.add(newInterval);

        // After new interval
        while (i < n) {
            result.add(intervals[i]);
            i++;
        }

        return result;
    }

    public static void main(String[] args) {
        int[][] intervals = {{1, 3}, {6, 9}};
        int[] newInterval = {2, 5};

        List<int[]> result = insert(intervals, newInterval);

        System.out.println("Updated intervals:");
        for (int[] interval : result) {
            System.out.println(Arrays.toString(interval));
        }
    }
}
